h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -580,27 +664,38 @@ protected void compute() {
         DKV.put(oneAugmentedColumnCenter);
         System.arraycopy(oneAugmentedColumnCenter.names(), 0, _gamColNamesCenter[_gamColIndex], 0,
                 _numKnotsM1);
-        copy2DArray(genOneGamCol._bInvD, _binvD[_csIndex]);
+        copy2DArray(genOneGamCol._bInvD, _binvD[_singlePredSplineInd]);
       }
     }
 
     void addGAM2Train() {
       final int numGamFrame = _parms._gam_columns.length; // number of smoothers to generate
       RecursiveAction[] generateGamColumn = new RecursiveAction[numGamFrame];
       int thinPlateInd = 0;
-      int csInd = 0;
+      int singlePredictorSmootherInd = 0;
       Frame trainFrame = _parms.train();
       for (int index = 0; index < numGamFrame; index++) { // generate smoothers/splines
         final Frame predictVec = prepareGamVec(index, _parms, trainFrame);// extract predictors from frame
-        final int numKnots = _parms._num_knots_sorted[index];
+        // numKnots for I-spline will be the number of basis
+        final int numKnots = _parms._bs_sorted[index] == 2 ? 
+                _parms._num_knots_sorted[index] + _parms._spline_orders_sorted[index] - 2 : 
+                _parms._num_knots_sorted[index];
         final int numKnotsM1 = numKnots - 1;
-        if (_parms._bs_sorted[index] == 0) {  // for CS smoothers
+        if (_parms._bs_sorted[index] == 0 || _parms._bs_sorted[index] == 2) {  // for CS or I-spline smoothers
           _gamColNames[index] = generateGamColNames(index, _parms);
-          _gamColNamesCenter[index] = new String[numKnotsM1];
+          if (_parms._bs_sorted[index]==2) 
+            _gamColNamesCenter[index] = new String[numKnots];
+          else
+            _gamColNamesCenter[index] = new String[numKnotsM1];
           _gamColMeans[index] = new double[numKnots];
-          generateGamColumn[index] = new CubicSplineSmoother(predictVec, _parms, index, _gamColNames[index],
-                  _knots[index][0], firstTwoLess, csInd++);
-        } else {  // TP splines with knots
+          if (_parms._bs_sorted[index] == 0) { // cs spline
+            generateGamColumn[index] = new CubicSplineSmoother(predictVec, _parms, index, _gamColNames[index],
+                    _knots[index][0], singlePredictorSmootherInd++);
+          } else { // I-splines
+            generateGamColumn[index] = new ISplineSmoother(predictVec, _parms, index, _gamColNames[index],
+                    _knots[index][0], singlePredictorSmootherInd++);
+          }
+        }  else {  // TP splines with knots
```

Comment:
seems that L686 - L698 could be simplified to just one if-else:

```
...
_gamColNames[index] = generateGamColNames(index, _parms);
_gamColMeans[index] = new double[numKnots];
if (_parms._bs_sorted[index]==2) {
    _gamColNamesCenter[index] = new String[numKnots];
    generateGamColumn[index] = new ISplineSmoother(predictVec, _parms, index, _gamColNames[index],
                     _knots[index][0], singlePredictorSmootherInd++);
} else {
    _gamColNamesCenter[index] = new String[numKnotsM1];
   generateGamColumn[index] = new CubicSplineSmoother(predictVec, _parms, index, _gamColNames[index],
                     _knots[index][0], singlePredictorSmootherInd++);
}
...
```

Reply:
You are right!  Thanks.